internal/plugin/connectors/http/generic/oauth/v1/protocol.go

```diff
@@ -0,0 +1,275 @@
+	for _, requiredParam := range requiredConfigParams {
+		if len(params[requiredParam]) < 1 {
+			return fmt.Errorf("required oAuth1 parameter '%s' not found", requiredParam)
+		}
+	}
+	return nil
+}
+
+/*
+extractOAuthParams extracts the required oAuth1 parameters from the config and assigns the other required paramters
+*/
+func extractOAuthParams(params map[string]string) oAuth1 {
+	oauth := oAuth1{
+		ConsumerKey:     params[consumerKey],
+		ConsumerSecret:  params[consumerSecret],
+		Nonce:           generateNonce(nonceLength, nonceCharset),
+		SignatureMethod: oauthSignatureMethodType,
+		Token:           params[token],
+		TokenSecret:     params[tokenSecret],
+		TimeStamp:       strconv.Itoa(int(time.Now().Unix())),
+		Version:         oauthVersionNumber,
+	}
+
+	return oauth
+}
+
+/*
+encodeURI percent encodes a string and changes encoding of <SPACE> from "+" to "%20"
+
+Reference: https://tools.ietf.org/html/rfc5849#section-3.6
+*/
+func encodeURI(stringToEncode string) string {
+	return strings.ReplaceAll(url.QueryEscape(stringToEncode), "+", "%20")
+}
+
+/*
+collectParameters creates a map of body and query parameters by percent encoding all
+keys and values and sorting values post-encoding
+*/
+func collectParameters(oauth1 oAuth1, r *gohttp.Request) map[string][]string {
+	// Create two copies of the request body
+	// Since r.Body is io.ReadCloser the body is empty after reading.
+	// Need to replace r.Body after r.ParseForm() dumps the body contents
+	buf, _ := ioutil.ReadAll(r.Body)
+	originalBody := ioutil.NopCloser(bytes.NewBuffer(buf))
+	copyBody := ioutil.NopCloser(bytes.NewBuffer(buf))
+	r.Body = originalBody
+
+	paramMap := make(map[string][]string, 0)
+	// get query and body params and URL encode keys and values sort values after encoding
+	r.ParseForm()
+	for paramKey, paramValues := range r.Form {
+		encodedKey := encodeURI(paramKey)
+		paramMap[encodedKey] = paramValues
+
+		// Keys with multiple values require the values to be sorted post-percent encoding
+		for valueIndex, value := range paramMap[encodedKey] {
+			encodedValue := encodeURI(value)
+			paramMap[encodedKey][valueIndex] = encodedValue
+		}
+		sort.Strings(paramMap[encodedKey])
+	}
+	r.Body = copyBody
+
+	// these keys are already URL safe and don't need sorted(1 value)
+	paramMap[oauthConsumerKey] = []string{encodeURI(oauth1.ConsumerKey)}
+	paramMap[oauthNonce] = []string{encodeURI(oauth1.Nonce)}
+	paramMap[oauthSignatureMethod] = []string{encodeURI(oauth1.SignatureMethod)}
+	paramMap[oauthTimestamp] = []string{encodeURI(oauth1.TimeStamp)}
+	paramMap[oauthToken] = []string{encodeURI(oauth1.Token)}
+	paramMap[oauthVersion] = []string{encodeURI(oauth1.Version)}
+
+	return paramMap
+}
+
+/*
+generateParameterString creates a Parameter String from sorting an encoded map of parameters and values
+
+Reference: https://tools.ietf.org/html/rfc5849#section-3.5
+*/
+func generateParameterString(paramMap map[string][]string) string {
+	var parameterString bytes.Buffer
```

Comment:
I wonder if maybe https://golang.org/pkg/strings/#Builder would be better here. https://medium.com/@felipedutratine/string-concatenation-in-golang-since-1-10-bytes-buffer-vs-strings-builder-2b3081848c45.

Reply:
lgtm! I hadn't heard of `strings.Builder` but it makes sense to use that instead